app/models/manageiq/providers/vmware/infra_manager/vm/reconfigure.rb

```diff
@@ -37,21 +37,132 @@ def max_memory_mb
       when "07"
         ec =  VimArray.new('ArrayOfOptionValue')
         ec << VimHash.new('OptionValue') do |ov|
           ov.key   = "cpuid.coresPerSocket"
-          ov.value = VimString.new(task_options[:cores_per_socket].to_s, nil, "xsd:string")
+          ov.value = VimString.new(options[:cores_per_socket].to_s, nil, "xsd:string")
         end
         vmcs.extraConfig = ec
       else
-        set_spec_option(vmcs, :numCoresPerSocket, task_options[:cores_per_socket], :to_i)
+        set_spec_option(vmcs, :numCoresPerSocket, options[:cores_per_socket], :to_i)
       end
-      set_spec_option(vmcs, :memoryMB, task_options[:vm_memory],      :to_i)
-      set_spec_option(vmcs, :numCPUs,  task_options[:number_of_cpus], :to_i)
+      set_spec_option(vmcs, :memoryMB, options[:vm_memory],      :to_i)
+      set_spec_option(vmcs, :numCPUs,  options[:number_of_cpus], :to_i)
+
+      if options[:disk_remove] || options[:disk_add]
+        with_provider_object do |vim_obj|
+          options[:disk_remove].each { |d| remove_disk_config_spec(vim_obj, vmcs, d) } if options[:disk_remove]
+          add_disks(vim_obj, vmcs, options[:disk_add]) if options[:disk_add]
+        end
+      end
+    end
+  end
+
+  def add_disks(vim_obj, vmcs, disks)
+    controller_key, unit_number = vim_obj.send(:getScsiCandU)
+
+    # if there is no scsi controller
+    if controller_key.nil?
+      controller_key, unit_number = [-99, 0]
+      add_scsi_controller(vmcs, 0, controller_key)
+    end
+
+    disks.each do |d|
+      d[:controller_key] = controller_key
+      d[:unit_number]    = unit_number
+      add_disk_config_spec(vmcs, d)
+      unit_number += 1
+    end
+  end
+
+  def add_scsi_controller(vmcs, bus_number, dev_key)
+    device_type = 'VirtualLsiLogicController'
+    add_device_config_spec(vmcs, VirtualDeviceConfigSpecOperation::Add) do |vdcs|
+      vdcs.device = VimHash.new(device_type) do |dev|
+        dev.sharedBus = VimString.new('noSharing', 'VirtualSCSISharing')
+        dev.busNumber = bus_number
+        dev.key       = dev_key
+      end
+    end
+  end
+
+  def backing_filename(name, unit_number)
+    filename = (name.blank? ? "disk_#{Time.zone.now.iso8601.gsub(/[-:TZ]/, '')}" : name) + "_#{unit_number}.vmdk"
+
+    # create the new disk in the same folder as the primary_diskfile or the VM's config file
+    primary_diskfile = hardware.disks.order(:location).find_by(:device_type => 'disk').try(:filename)
+    new_disk_path    = File.dirname(primary_diskfile || path)
+    File.join(new_disk_path, filename)
+  end
+
+  def disk_mode(dependent, persistent)
+    if dependent
+      persistent ? VirtualDiskMode::Persistent : VirtualDiskMode::Nonpersistent
+    else
+      persistent ? VirtualDiskMode::Independent_persistent : VirtualDiskMode::Independent_nonpersistent
+    end
+  end
+
+  def add_disk_config_spec(vmcs, options)
+    raise "#{__method__}: Disk size is required to add a new disk." unless options[:disk_size]
```

Comment:
@lfu We should explicitly state the unit used for `:disk_size`.  Maybe `:disk_size_in_mb`